### PR TITLE
Solve early memory bootstrap and remove need for heap

### PR DESCRIPTION
**Pull request by n1tram1**
Previously we have an early bootstrap allocator than supposed to be used later for a more complete page allocator. We also had a fixed size heap for said bootstrap allocator.

Now we don't need any of that, we just initialize a proper page allocator from the device tree.

### CODE REVIEW

**Comment by n1tram1** [comment]
Oops sorry, I forgot to remove the `SimplePageAllocator`

File: src/mm/page_manager.rs [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    page_size: usize,
+}
+
+impl<'a> PageManager<'a> {
+
+    fn count_pages(regions: impl Iterator<Item = fdt::standard_nodes::MemoryRegion>, page_size: usize) -> usize {
+        regions
+            .map(|region| {
+                (region.starting_address, unsafe {
+                    region.starting_address.add(region.size.unwrap_or(0))
```
Comment (Skallwar) [inline src/mm/page_manager.rs:63]: Could we not unwrap only. If there is an error here, we should not continue any further

File: src/mm/page_manager.rs [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+        let mut first_page_addr: usize = 0;
+        let mut consecutive_pages: usize = 0;
+
+        for page in physical_pages {
+            if consecutive_pages == 0 {
+                first_page_addr = page.base;
+            }
+
+            if !page.is_unused() {
```
Comment (Skallwar) [inline src/mm/page_manager.rs:115]:
```suggestion
            if page.is_used() {
```
I find this more readable

File: src/mm/page_manager.rs [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+        None
+    }
+
+    /// TLDR: Initialize a [`PageAllocator`] from the device tree.
+    /// How it works
+    /// First count how many pages we can make out on the system, how much size we will need for
+    /// metadata and align that up to a page size.
+    /// Second (in [`Self::find_contiguous_unused_pages`]), look through our pages for a contiguous
+    /// space large enough to hold all our metadata.
+    /// Lastly store our metadata there, and mark pages as unused or kernel.
```
Comment (Skallwar) [inline src/mm/page_manager.rs:136]:
```suggestion
    /// How it works:
    /// - First count how many pages we can make out on the system, how much size we will need for
    /// metadata and align that up to a page size.
    /// - Second (in [`Self::find_contiguous_unused_pages`]), look through our pages for a contiguous
    /// space large enough to hold all our metadata.
    /// - Lastly store our metadata there, and mark pages as unused or kernel.
```

File: src/arch/mod.rs [hunk 4]
```diff
@@ -23,17 +23,17 @@ pub fn new_arch() -> impl Architecture {
 }
 
 pub trait Architecture {
-    unsafe extern "C" fn _start() -> !;
+    unsafe extern "C" fn _start() -> ();
```
[the remark this answers is not in the document]
Comment (n1tram1) [reply src/arch/mod.rs:26]: fixed

File: src/mm/page_manager.rs (hunk 1 above)
Comment (n1tram1) [reply src/mm/page_manager.rs:64]: It's not exactly an error.
Further regions could have a proper size.

File: src/mm/page_alloc.rs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+#[derive(Debug)]
+pub enum AllocError {
```
[the remark this answers is not in the document]
Comment (n1tram1) [reply src/mm/page_alloc.rs]: fixed, by de9279fd59365af326422764c3c94ac628fefbcb

File: src/main.rs [hunk 6]
```diff
@@ -21,8 +20,10 @@ use drivers::plic;
 
 use arch::Architecture;
 
+use core::arch::asm;
+
 #[no_mangle]
-fn k_main() -> ! {
+fn k_main(_core_id: usize, device_tree_ptr: usize) -> ! {
```
[the remark this answers is not in the document]
Comment (n1tram1) [reply src/main.rs:26]: fixed fc9564f

File: src/arch/mod.rs [hunk 7]
```diff
@@ -29,11 +29,11 @@ pub trait Architecture {
 }
 
 pub trait ArchitectureMemory {
-    fn new<'alloc>(allocator: &mut mm::SimplePageAllocator<'alloc>) -> &'alloc mut Self;
+    fn new<'alloc>(allocator: &mut impl mm::PageAllocator) -> &'alloc mut Self;
```
Comment (CohenArthur) [inline src/arch/mod.rs:32]:
```suggestion
    fn new<'alloc>(allocator: &mut dyn mm::PageAllocator) -> &'alloc mut Self;
```
I think this is more idiomatic. References on trait objects should use `dyn`, while actual moved trait objects should use `impl`.

File: src/mm/mod.rs [hunk 8]
```diff
@@ -23,46 +22,80 @@ bitflags! {
...
+        (
+            utils::external_symbol_value(&KERNEL_START),
+            utils::external_symbol_value(&KERNEL_END),
+        )
+    };
+
+    base >= kernel_start && base < kernel_end
+}
+
+// TODO: shall this be moved to arch/riscv (in the MemoryImpl) ?
```
[the remark this answers is not in the document]
Comment (CohenArthur) [reply src/mm/mod.rs:35]: Agreed, it's better to keep things as architecture agnostic as possible in these common files imo

**Review by CohenArthur — APPROVED** [review]

**Review by Skallwar — CHANGES_REQUESTED** [review]
There is still some comments that needs to be addresesd.

We really need to put some documentation about why we are doing such a hack withe the lambda with the iterator in the `Architecture` trait

File: src/arch/mod.rs (hunk 7 above)
Comment (n1tram1) [reply src/arch/mod.rs:38]: cannot be done
```rust
error[E0038]: the trait `Architecture` cannot be made into an object
  --> src/mm/mod.rs:55:23
   |
55 |     pub fn new(arch: &dyn arch::Architecture) -> Self {
   |                       ^^^^^^^^^^^^^^^^^^^^^^ `Architecture` cannot be made into an object
   |
note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
```
And this is only the first 20% of the error message

**Review by Skallwar — COMMENTED** [review]
Need some documentation but I'm happy with that

**Review by CohenArthur — APPROVED** [review]
LGTM

**Review by CohenArthur — APPROVED** [review]

**Pull request merged**